contract/models/contract_line.py

```diff
@@ -513,30 +513,44 @@ def _prepare_invoice_line(self, invoice_id=False):
         return invoice_line_vals
 
     @api.multi
-    def _get_invoiced_period(self):
+    def _get_next_invoiced_period(
+        self, last_date_invoiced, recurring_next_date
+    ):
         self.ensure_one()
         first_date_invoiced = (
-            self.last_date_invoiced + relativedelta(days=1)
-            if self.last_date_invoiced
+            last_date_invoiced + relativedelta(days=1)
+            if last_date_invoiced
             else self.date_start
         )
         if self.recurring_rule_type == 'monthlylastday':
-            last_date_invoiced = self.recurring_next_date
+            last_date_invoiced = recurring_next_date
         else:
             if self.recurring_invoicing_type == 'pre-paid':
                 last_date_invoiced = (
-                    self.recurring_next_date
+                    recurring_next_date
                     + self.get_relative_delta(
                         self.recurring_rule_type, self.recurring_interval
                     )
                     - relativedelta(days=1)
                 )
             else:
-                last_date_invoiced = self.recurring_next_date - relativedelta(
+                last_date_invoiced = recurring_next_date - relativedelta(
                     days=1
                 )
         if self.date_end and self.date_end < last_date_invoiced:
             last_date_invoiced = self.date_end
+        recurring_next_date = recurring_next_date + self.get_relative_delta(
+            self.recurring_rule_type, self.recurring_interval
+        )
+        return first_date_invoiced, last_date_invoiced, recurring_next_date
+
+    @api.multi
+    def _get_invoiced_period(self):
+        self.ensure_one()
+        dates = self._get_next_invoiced_period(
+            self.last_date_invoiced, self.recurring_next_date
+        )
+        first_date_invoiced, last_date_invoiced, recurring_next_date = dates
         return first_date_invoiced, last_date_invoiced
```

Comment:
@sbejaoui another question: it's looks strange that `recurring_next_date` is computed by `_get_next_invoiced_period` then ignored in the return here, and computed (differently) elsewhere in `_update_recurring_next_date`. Also this `_get_invoiced_period` now looks a little bit useless. 

Perhaps you can push the refactoring further?

Reply:
yes, I was in a hurry to get out `_get_next_invoiced_period` to use it for the forecast algorithm. Unify the invoiced period compute is in my to do list.